File: src/day06/main.exs
defmodule Day06 do
  def all_uniq(s) do
    String.graphemes(s)
    |> Enum.uniq()
    |> Enum.count() ==
      String.length(s)
  end

  def detect(input, n) do
    {_, i} =
      Day06.windows(input, n)
      |> Stream.with_index()
      |> Stream.filter(fn {s, _} ->
        Day06.all_uniq(s)
      end)
      |> Enum.at(0)

    i + n
  end

  def windows(s, n) do
    n..String.length(s)
    |> Stream.map(fn i ->
      String.slice(s, (i - n)..(i - 1))
    end)
  end
end

input =
  File.read!("puzzle_input/day06")
  |> String.trim_trailing()

IO.puts("part 1: #{Day06.detect(input, 4)}")
IO.puts("part 2: #{Day06.detect(input, 14)}")

File: src/day06/main.hs
import Data.List (foldl')

allUniq :: (Eq a) => [a] -> Bool
allUniq [] = True
allUniq (x : xs) = x `notElem` xs && allUniq xs

fst3 :: (a, b, c) -> a
fst3 (x, _, _) = x

windows :: Int -> String -> [String]
windows n s =
  fst3 $ foldl' f ([], 0, "") s
  where
    f :: ([String], Int, String) -> Char -> ([String], Int, String)
    f (wins, len, prev) c
      | len < n = (wins, len + 1, prev ++ [c])
      | otherwise = (wins ++ [prev], len, drop 1 prev ++ [c])

detect n s =
  case filter (allUniq . snd) $ zip [0 ..] $ windows n s of
    ((i, _) : _) -> i
    _ -> error "empty list"

main = do
  input <- readFile "puzzle_input/day06"
  putStrLn $ (++) "part 1: " $ show $ detect 4 input
  putStrLn $ (++) "part 2: " $ show $ detect 14 input